### PR TITLE
Index.writeBasedOnFeaturePath should throw instead of silently Failing

### PR DESCRIPTION
**Pull request by magicDGS**
### Description

Solves #821

### Checklist

- [x] Code compiles correctly
- [ ] New tests covering changes and new functionality
- [x] All tests passing
- [ ] Extended the README / documentation, if necessary
- [x] Is not backward compatible (breaks binary or source compatibility)

### CODE REVIEW

**Comment by magicDGS** [comment]
Here you have a patch that should be tested against picard/gatk, @lbergelson. Silently failing is not nice...

File: src/main/java/htsjdk/tribble/index/tabix/TabixIndex.java [hunk 1]
```diff
@@ -230,8 +228,7 @@ public void write(final Path tabixPath) throws IOException {
     @Override
     public void writeBasedOnFeaturePath(final Path featurePath) throws IOException {
```
Comment (droazen) [inline src/main/java/htsjdk/tribble/index/tabix/TabixIndex.java]: Can you update the javadoc for this method in the `Index` interface to mention that this method now throws an `IOException` in the case where `featurePath` is not a regular file?

File: src/main/java/htsjdk/tribble/index/tabix/TabixIndex.java [hunk 2]
```diff
@@ -230,8 +228,7 @@ public void write(final Path tabixPath) throws IOException {
...
         if (!Files.isRegularFile(featurePath)) {
-            LOGGER.warn("Index not written into ", featurePath);
-            return;
+            throw new IOException("Cannot write based on a non-regular file: " + featurePath.toUri());
```
Comment (droazen) [inline src/main/java/htsjdk/tribble/index/tabix/TabixIndex.java]: Can you add a test case showing that the new exception is thrown?

**Comment by droazen** [comment]
Two comments, back to @magicDGS for changes

File: src/main/java/htsjdk/tribble/index/tabix/TabixIndex.java (hunk 1 above)
Comment (lbergelson) [inline src/main/java/htsjdk/tribble/index/tabix/TabixIndex.java]: also update the javadoc for the associated file methods

Comment (magicDGS) [inline src/main/java/htsjdk/tribble/index/tabix/TabixIndex.java]: Done.

File: src/main/java/htsjdk/tribble/index/tabix/TabixIndex.java (hunk 2 above)
Comment (magicDGS) [inline src/main/java/htsjdk/tribble/index/tabix/TabixIndex.java]: Done.

**Comment by magicDGS** [comment]
I've addressed the comments - back to you @droazen. Thanks for reviewing!

**Comment by lbergelson** [comment]
@magicDGS The changes here look good.  @yfarjoun wants to run this against the picard tests to make sure nothing was relying on the old behavior.

**Comment by yfarjoun** [comment]
not entirely sure what's going on, but when I build picard with this htsjdk I get the following error:

```
htsjdk.samtools.SAMException: Missing sequence header at line 1 in fastq /Users/farjoun/picard-private/Picard-public/testdata/picard/sam/fastq2bam/permissive-format/s_1_2_sequence.txt
	at htsjdk.samtools.fastq.FastqReader.readNextRecord(FastqReader.java:107)
	at htsjdk.samtools.fastq.FastqReader.<init>(FastqReader.java:93)
	at htsjdk.samtools.fastq.FastqReader.<init>(FastqReader.java:77)
	at picard.sam.FastqToSam.fileToFastqReader(FastqToSam.java:360)
	at picard.sam.FastqToSam.doWork(FastqToSam.java:251)
	at picard.cmdline.CommandLineProgram.instanceMain(CommandLineProgram.java:205)
	at picard.cmdline.PicardCommandLine.instanceMain(PicardCommandLine.java:94)
	at picard.cmdline.PicardCommandLine.instanceMain(PicardCommandLine.java:99)
```

**Comment by yfarjoun** [comment]
looks like the problem goes away with a rebase onto htsjdk/master, so nevermind.

**Comment by yfarjoun** [comment]
👍

**Comment by magicDGS** [comment]
Thnaks! And sorry for the inconvenience of the rebasing issue!

**Pull request merged**